src/openms/source/QC/MZcalibration.cpp

```diff
@@ -0,0 +1,115 @@
+// --------------------------------------------------------------------------
+
+
+#include <OpenMS/QC/QCBase.h>
+#include <OpenMS/KERNEL/FeatureMap.h>
+#include <OpenMS/QC/MzCalibration.h>
+#include <OpenMS/KERNEL/MSExperiment.h>
+#include <OpenMS/FORMAT/IdXMLFile.h>
+#include <OpenMS/CONCEPT/Types.h>
+#include <OpenMS/CONCEPT/Exception.h>
+#include <algorithm>
+
+using namespace std;
+
+namespace OpenMS
+{
+		// find original m/z Value, set meta value "mz_raw" and set meta value "mz_ref"
+		void MzCalibration::compute(FeatureMap& features, const MSExperiment& exp)
+		{
+				if (features.empty())
+				{
+						LOG_WARN << "The FeatureMap is empty.\n";
+				}
+				if (exp.empty())
+				{
+						throw Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "The PeakMap is empty.");
+				}
+
+				//set meta values for the first hit of all PeptideIdentifications of all features
+				for (Feature& feature : features)
+				{
+						for (PeptideIdentification& peptide_ID : feature.getPeptideIdentifications())
+						{
+								if (peptide_ID.hasRT())
+								{
+										peptide_ID.getHits()[0].setMetaValue("mz_raw", getMZraw_(peptide_ID.getRT(), exp));
+										peptide_ID.getHits()[0].setMetaValue("mz_ref", (peptide_ID.getHits()[0].getSequence().getMonoWeight(OpenMS::Residue::Full, peptide_ID.getHits()[0].getCharge())) / peptide_ID.getHits()[0].getCharge());
+								}
+						}
+				}
+				//set meta values for the first hit of all unasssigned PeptideIdentifications
+				for (PeptideIdentification& unassigned_ID : features.getUnassignedPeptideIdentifications())
+				{
+						if (unassigned_ID.hasRT())
+						{
+								unassigned_ID.getHits()[0].setMetaValue("mz_raw", getMZraw_(unassigned_ID.getRT(), exp));
+								unassigned_ID.getHits()[0].setMetaValue("mz_ref", (unassigned_ID.getHits()[0].getSequence().getMonoWeight(OpenMS::Residue::Full, unassigned_ID.getHits()[0].getCharge())) / unassigned_ID.getHits()[0].getCharge());
+						}
+				}
+		}
+		//required input files
+		QCBase::Status MzCalibration::requires() const
+		{
+				return QCBase::Status() | QCBase::Requires::RAWMZML | QCBase::Requires::POSTFDRFEAT;
+		}
+
+		// search matching RT-time in MSExperiment before calibration, and return the m/z value
+		double MzCalibration::getMZraw_(double rt, const MSExperiment& exp) const
+		{
+				MSExperiment::ConstIterator it = exp.RTBegin(rt - EPSILON_);
+				
+				if (it == exp.end())
+				{
+						throw Exception::IllegalArgument(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "The retention time of the MZML and featureXML file does not match.");
+				}
+
+				const auto& spectrum = *it;
+				
+				if (spectrum.getRT() - rt > EPSILON_)
+				{ 
+						throw Exception::IllegalArgument(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "The retention time of the MZML and featureXML file does not match.");
+				}
+			
+				if (spectrum.getMSLevel() == 2)
+				{
+						return spectrum.getPrecursors()[0].getMetaValue("mz_raw");
```

Comment:
check `...hasMetaValue("mz_raw")) and throw Exception  "InternalCalibration was not called; MetaValue 'mz_raw' missing from MSExperiment."